File: Linked List/leetcode_19.py

# Problem Link ====>>https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/


class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None


class Solution:
    def removeNthFromEnd(self, head, n):
        dummy = ListNode(0)
        dummy.next = head
        first = dummy
        second = dummy

        for i in range(1, n + 1):
            first = first.next

        while first:
            first = first.next
            second = second.next

        second.next = second.next.next
        return dummy.next


def main():
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)

    solution = Solution()
    n = 2
    newHead = solution.removeNthFromEnd(head, n)

    while newHead:
        print(newHead.val, end=" ")
        newHead = newHead.next

    print()

    del head
    del newHead


if __name__ == "__main__":
    main()

File: Linked List/leetcode_19.cpp

// Problem Link ====>>https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/

#include <iostream>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
    public:
        ListNode *removeNthFromEnd(ListNode *head, int n)
        {
            ListNode *dummy = new ListNode(0);
            dummy->next = head;
            ListNode *first = dummy;
            ListNode *second = dummy;

            for (int i = 1; i <= n + 1; ++i)
            {
                first = first->next;
            }

            while (first != nullptr)
            {
                first = first->next;
                second = second->next;
            }

            second->next = second->next->next;
            return dummy->next;
        }
};

int main()
{
    // Create the linked list
    ListNode *head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = new ListNode(5);

    // Create a Solution object
    Solution solution;

    // Remove the nth node from the end
    int n = 2;
    ListNode *newHead = solution.removeNthFromEnd(head, n);

    // Print the new linked list
    while (newHead != nullptr)
    {
        cout << newHead->val << " ";
        newHead = newHead->next;
    }

    cout << endl;

    // Delete the linked list
    delete head;
    delete newHead;

    return 0;
}